File: tests/python/test_sama.py
"""
Python binding tests for SAMA indicator.
These tests mirror the Rust unit tests to ensure Python bindings work correctly.
"""
import pytest
import numpy as np
import sys
from pathlib import Path

try:
    import my_project as ta_indicators
except ImportError:

    try:
        import my_project as ta_indicators
    except ImportError:
        pytest.skip("Python module not built. Run 'maturin develop --features python' first", allow_module_level=True)

from test_utils import load_test_data, assert_close, EXPECTED_OUTPUTS
from rust_comparison import compare_with_rust


class TestSama:
    @pytest.fixture(scope='class')
    def test_data(self):
        return load_test_data()

    def test_sama_partial_params(self, test_data):
        """Test SAMA with partial parameters - mirrors check_sama_partial_params"""
        close = test_data['close']


        result = ta_indicators.sama(close, 200, 14, 6)
        assert len(result) == len(close)


        result2 = ta_indicators.sama(close, 50, 14, 6)
        assert len(result2) == len(close)

    def test_sama_accuracy(self, test_data):
        """Test SAMA matches expected values from Rust tests - mirrors check_sama_accuracy"""
        close = test_data['close'][:300]
        expected = EXPECTED_OUTPUTS['sama']


        result = ta_indicators.sama(
            close,
            length=expected['default_params']['length'],
            maj_length=expected['default_params']['maj_length'],
            min_length=expected['default_params']['min_length']
        )

        assert len(result) == len(close)


        valid_values = result[~np.isnan(result)]
        if len(valid_values) >= 5:
            assert_close(
                valid_values[-5:],
                expected['last_5_values'],
                rtol=1e-8,
                msg="SAMA last 5 values mismatch (default params)"
            )


        result2 = ta_indicators.sama(
            close,
            length=expected['test_params']['length'],
            maj_length=expected['test_params']['maj_length'],
            min_length=expected['test_params']['min_length']
        )

        valid_values2 = result2[~np.isnan(result2)]
        assert len(valid_values2) >= 5, "Should have at least 5 valid values with length=50"
        assert_close(
            valid_values2[-5:],
            expected['test_last_5'],
            rtol=1e-8,
            msg="SAMA last 5 values mismatch (test params)"
        )

    def test_sama_default_candles(self, test_data):
        """Test SAMA with default parameters - mirrors check_sama_default_candles"""
        close = test_data['close']


        result = ta_indicators.sama(close, 200, 14, 6)
        assert len(result) == len(close)



        assert not np.all(np.isnan(result[:200])), "Should have computed values, not all NaN"

    def test_sama_zero_period(self):
        """Test SAMA fails with zero period - mirrors check_sama_zero_period"""
        input_data = np.array([10.0, 20.0, 30.0])


        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.sama(input_data, length=0, maj_length=14, min_length=6)


        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.sama(input_data, length=10, maj_length=0, min_length=6)


        with pytest.raises(ValueError, match="Invalid period"):
            ta_indicators.sama(input_data, length=10, maj_length=14, min_length=0)

    def test_sama_period_exceeds_length(self):
        """Test SAMA fails when length exceeds data length - mirrors check_sama_period_exceeds_length"""
        data_small = np.array([10.0, 20.0, 30.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.sama(data_small, length=10, maj_length=14, min_length=6)

    def test_sama_very_small_dataset(self):
        """Test SAMA fails with insufficient data - mirrors check_sama_very_small_dataset"""
        single_point = np.array([42.0])

        with pytest.raises(ValueError, match="Invalid period|Not enough valid data"):
            ta_indicators.sama(single_point, length=200, maj_length=14, min_length=6)

    def test_sama_empty_input(self):
        """Test SAMA fails with empty input - mirrors check_sama_empty_input"""
        empty = np.array([])

        with pytest.raises(ValueError, match="Input data slice is empty"):
            ta_indicators.sama(empty, length=200, maj_length=14, min_length=6)

    def test_sama_all_nan(self):
        """Test SAMA fails with all NaN values - mirrors check_sama_all_nan"""
        all_nan = np.full(100, np.nan)

        with pytest.raises(ValueError, match="All values are NaN"):
            ta_indicators.sama(all_nan, length=50, maj_length=14, min_length=6)

    def test_sama_reinput(self, test_data):
        """Test SAMA applied twice (re-input) - mirrors check_sama_reinput"""
        close = test_data['close'][:300]
        expected = EXPECTED_OUTPUTS['sama']


        first_result = ta_indicators.sama(
            close,
            length=expected['test_params']['length'],
            maj_length=expected['test_params']['maj_length'],
            min_length=expected['test_params']['min_length']
        )
        assert len(first_result) == len(close)


        second_result = ta_indicators.sama(
            first_result,
            length=expected['test_params']['length'],
            maj_length=expected['test_params']['maj_length'],
            min_length=expected['test_params']['min_length']
        )
        assert len(second_result) == len(first_result)


        valid_reinput = second_result[~np.isnan(second_result)]
        assert len(valid_reinput) >= 5, "Should have at least 5 valid reinput values"
        assert_close(
            valid_reinput[-5:],
            expected['reinput_last_5'],
            rtol=1e-8,
            msg="SAMA re-input last 5 values mismatch"
        )

    def test_sama_nan_handling(self, test_data):
        """Test SAMA handles NaN values correctly - mirrors check_sama_nan_handling"""
        close = test_data['close']


        result = ta_indicators.sama(close, length=50, maj_length=14, min_length=6)
        assert len(result) == len(close)



        first_valid = np.where(~np.isnan(close))[0]
        if len(first_valid) > 0:
            first_valid_idx = first_valid[0]


            if first_valid_idx < len(result) - 10:

                initial_values = result[first_valid_idx:first_valid_idx+10]
                assert not np.all(np.isnan(initial_values)), "Should have valid values from first valid input"

    def test_sama_streaming(self, test_data):
        """Test SAMA streaming matches batch calculation - mirrors check_sama_streaming"""
        close = test_data['close'][:300]
        length = 50
        maj_length = 14
        min_length = 6


        batch_result = ta_indicators.sama(close, length=length, maj_length=maj_length, min_length=min_length)


        stream = ta_indicators.SamaStream(length=length, maj_length=maj_length, min_length=min_length)
        stream_values = []

        for price in close:
            result = stream.update(price)
            stream_values.append(result if result is not None else np.nan)

        stream_values = np.array(stream_values)


        assert len(batch_result) == len(stream_values)


        for i, (b, s) in enumerate(zip(batch_result, stream_values)):
            if np.isnan(b) and np.isnan(s):
                continue
            assert_close(b, s, rtol=1e-9, atol=1e-9,
                        msg=f"SAMA streaming mismatch at index {i}")

    def test_sama_batch_single_params(self, test_data):
        """Test SAMA batch processing with single parameter set - mirrors check_batch_default_row"""
        close = test_data['close'][:300]
        expected = EXPECTED_OUTPUTS['sama']

        result = ta_indicators.sama_batch(
            close,
            length_range=(200, 200, 0),
            maj_length_range=(14, 14, 0),
            min_length_range=(6, 6, 0)
        )

        assert 'values' in result
        assert 'lengths' in result
        assert 'maj_lengths' in result
        assert 'min_lengths' in result


        assert result['values'].shape[0] == 1
        assert result['values'].shape[1] == len(close)


        default_row = result['values'][0]
        valid_values = default_row[~np.isnan(default_row)]


        if len(valid_values) >= 5:
            assert_close(
                valid_values[-5:],
                expected['last_5_values'],
                rtol=1e-8,
                msg="SAMA batch default row mismatch"
            )

    def test_sama_batch_sweep(self, test_data):
        """Test SAMA batch processing with parameter sweep - mirrors check_batch_sweep"""
        close = test_data['close'][:100]

        result = ta_indicators.sama_batch(
            close,
            length_range=(40, 50, 5),
            maj_length_range=(12, 14, 1),
            min_length_range=(4, 6, 1)
        )


        expected_count = 3 * 3 * 3
        assert result['values'].shape[0] == expected_count, f"Expected {expected_count} batch results"
        assert result['values'].shape[1] == len(close), "Should have columns equal to data length"


        assert len(result['lengths']) == expected_count
        assert len(result['maj_lengths']) == expected_count
        assert len(result['min_lengths']) == expected_count


        assert result['lengths'][0] == 40
        assert result['maj_lengths'][0] == 12
        assert result['min_lengths'][0] == 4


        assert result['lengths'][-1] == 50
        assert result['maj_lengths'][-1] == 14
        assert result['min_lengths'][-1] == 6

    def test_sama_batch_vs_single(self, test_data):
        """Test batch result matches single calculation for same parameters"""
        close = test_data['close'][:100]


        single_result = ta_indicators.sama(close, length=45, maj_length=13, min_length=5)


        batch_result = ta_indicators.sama_batch(
            close,
            length_range=(45, 45, 0),
            maj_length_range=(13, 13, 0),
            min_length_range=(5, 5, 0)
        )


        assert_close(
            batch_result['values'][0],
            single_result,
            rtol=1e-10,
            msg="Batch vs single calculation mismatch"
        )


if __name__ == '__main__':
    pytest.main([__file__, '-v'])

File: tests/wasm/test_sama.js

import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import {
    loadTestData,
    assertArrayClose,
    assertClose,
    isNaN,
    assertAllNaN,
    assertNoNaN,
    EXPECTED_OUTPUTS
} from './test_utils.js';
import { compareWithRust } from './rust-comparison.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

test.before(async () => {

    try {
        const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
        const importPath = process.platform === 'win32'
            ? 'file:///' + wasmPath.replace(/\\/g, '/')
            : wasmPath;
        wasm = await import(importPath);

    } catch (error) {
        console.error('Failed to load WASM module. Run "wasm-pack build --features wasm --target nodejs" first');
        throw error;
    }

    testData = loadTestData();
});

test('SAMA partial params', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.sama_js(close, 200, 14, 6);
    assert.strictEqual(result.length, close.length);


    const result2 = wasm.sama_js(close, 50, 14, 6);
    assert.strictEqual(result2.length, close.length);
});

test('SAMA accuracy', async () => {

    const close = new Float64Array(testData.close.slice(0, 300));
    const expected = EXPECTED_OUTPUTS.sama;


    const result = wasm.sama_js(
        close,
        expected.defaultParams.length,
        expected.defaultParams.majLength,
        expected.defaultParams.minLength
    );

    assert.strictEqual(result.length, close.length);


    const validValues = result.filter(v => !isNaN(v));
    if (validValues.length >= 5) {
        const last5 = validValues.slice(-5);
        assertArrayClose(
            last5,
            expected.last5Values,
            1e-8,
            "SAMA last 5 values mismatch (default params)"
        );
    }


    const result2 = wasm.sama_js(
        close,
        expected.testParams.length,
        expected.testParams.majLength,
        expected.testParams.minLength
    );

    const validValues2 = result2.filter(v => !isNaN(v));
    assert(validValues2.length >= 5, "Should have at least 5 valid values with length=50");
    const last5Test = validValues2.slice(-5);
    assertArrayClose(
        last5Test,
        expected.testLast5,
        1e-8,
        "SAMA last 5 values mismatch (test params)"
    );
});

test('SAMA default candles', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.sama_js(close, 200, 14, 6);
    assert.strictEqual(result.length, close.length);



    assertNoNaN(result.slice(0, 200), "Should have valid values from start");
});

test('SAMA zero period', () => {

    const inputData = new Float64Array([10.0, 20.0, 30.0]);


    assert.throws(() => {
        wasm.sama_js(inputData, 0, 14, 6);
    }, /Invalid period/);


    assert.throws(() => {
        wasm.sama_js(inputData, 10, 0, 6);
    }, /Invalid period/);


    assert.throws(() => {
        wasm.sama_js(inputData, 10, 14, 0);
    }, /Invalid period/);
});

test('SAMA period exceeds length', () => {

    const dataSmall = new Float64Array([10.0, 20.0, 30.0]);

    assert.throws(() => {
        wasm.sama_js(dataSmall, 10, 14, 6);
    }, /Invalid period|Not enough valid data/);
});

test('SAMA very small dataset', () => {

    const singlePoint = new Float64Array([42.0]);

    assert.throws(() => {
        wasm.sama_js(singlePoint, 200, 14, 6);
    }, /Invalid period|Not enough valid data/);
});

test('SAMA empty input', () => {

    const empty = new Float64Array([]);

    assert.throws(() => {
        wasm.sama_js(empty, 200, 14, 6);
    }, /Input data slice is empty/);
});

test('SAMA all NaN input', () => {

    const allNaN = new Float64Array(100);
    allNaN.fill(NaN);

    assert.throws(() => {
        wasm.sama_js(allNaN, 50, 14, 6);
    }, /All values are NaN/);
});

test('SAMA reinput', () => {

    const close = new Float64Array(testData.close.slice(0, 300));
    const expected = EXPECTED_OUTPUTS.sama;


    const firstResult = wasm.sama_js(
        close,
        expected.testParams.length,
        expected.testParams.majLength,
        expected.testParams.minLength
    );
    assert.strictEqual(firstResult.length, close.length);


    const secondResult = wasm.sama_js(
        firstResult,
        expected.testParams.length,
        expected.testParams.majLength,
        expected.testParams.minLength
    );
    assert.strictEqual(secondResult.length, firstResult.length);


    const validReinput = secondResult.filter(v => !isNaN(v));
    assert(validReinput.length >= 5, "Should have at least 5 valid reinput values");
    const last5Reinput = validReinput.slice(-5);
    assertArrayClose(
        last5Reinput,
        expected.reinputLast5,
        1e-8,
        "SAMA re-input last 5 values mismatch"
    );
});

test('SAMA NaN handling', () => {

    const close = new Float64Array(testData.close);


    const result = wasm.sama_js(close, 50, 14, 6);
    assert.strictEqual(result.length, close.length);


    let firstValid = -1;
    for (let i = 0; i < close.length; i++) {
        if (!isNaN(close[i])) {
            firstValid = i;
            break;
        }
    }

    if (firstValid >= 0) {
        const warmupPeriod = firstValid + 50;


        if (warmupPeriod <= result.length) {

            assertNoNaN(result.slice(0, Math.min(10, result.length)), "Should have valid values from start");


            if (warmupPeriod < result.length - 10) {

                const afterWarmup = result.slice(warmupPeriod, warmupPeriod + 10);
                let hasValidValues = false;
                for (let v of afterWarmup) {
                    if (!isNaN(v)) {
                        hasValidValues = true;
                        break;
                    }
                }
                assert(hasValidValues, "Should have some valid values after warmup");
            }
        }
    }
});

test('SAMA batch single parameter set', () => {

    const close = new Float64Array(testData.close.slice(0, 300));
    const expected = EXPECTED_OUTPUTS.sama;


    const batchResult = wasm.sama_batch(close, {
        length_range: [200, 200, 0],
        maj_length_range: [14, 14, 0],
        min_length_range: [6, 6, 0]
    });


    assert(batchResult.values, 'Should have values array');
    assert(batchResult.combos, 'Should have combos array');
    assert(typeof batchResult.rows === 'number', 'Should have rows count');
    assert(typeof batchResult.cols === 'number', 'Should have cols count');


    assert.strictEqual(batchResult.rows, 1);
    assert.strictEqual(batchResult.cols, close.length);
    assert.strictEqual(batchResult.combos.length, 1);
    assert.strictEqual(batchResult.values.length, close.length);


    const combo = batchResult.combos[0];
    assert.strictEqual(combo.length, 200);
    assert.strictEqual(combo.maj_length, 14);
    assert.strictEqual(combo.min_length, 6);


    const validValues = batchResult.values.filter(v => !isNaN(v));
    if (validValues.length >= 5) {
        const last5 = validValues.slice(-5);
        assertArrayClose(
            last5,
            expected.last5Values,
            1e-8,
            "SAMA batch default row mismatch"
        );
    }
});

test('SAMA batch multiple parameters', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const batchResult = wasm.sama_batch(close, {
        length_range: [40, 50, 5],
        maj_length_range: [12, 14, 1],
        min_length_range: [4, 6, 1]
    });


    assert.strictEqual(batchResult.rows, 27);
    assert.strictEqual(batchResult.cols, 100);
    assert.strictEqual(batchResult.combos.length, 27);
    assert.strictEqual(batchResult.values.length, 27 * 100);


    assert.strictEqual(batchResult.combos[0].length, 40);
    assert.strictEqual(batchResult.combos[0].maj_length, 12);
    assert.strictEqual(batchResult.combos[0].min_length, 4);


    assert.strictEqual(batchResult.combos[26].length, 50);
    assert.strictEqual(batchResult.combos[26].maj_length, 14);
    assert.strictEqual(batchResult.combos[26].min_length, 6);


    for (let i = 0; i < Math.min(3, batchResult.combos.length); i++) {
        const combo = batchResult.combos[i];
        const rowStart = i * 100;
        const rowEnd = rowStart + 100;
        const rowData = batchResult.values.slice(rowStart, rowEnd);

        const singleResult = wasm.sama_js(close, combo.length, combo.maj_length, combo.min_length);
        assertArrayClose(
            rowData,
            singleResult,
            1e-10,
            `Row ${i} (length=${combo.length}, maj=${combo.maj_length}, min=${combo.min_length}) mismatch`
        );
    }
});

test('SAMA batch metadata from result', () => {

    const close = new Float64Array(60);
    close.fill(100);

    const result = wasm.sama_batch(close, {
        length_range: [40, 50, 10],
        maj_length_range: [12, 14, 2],
        min_length_range: [5, 6, 1]
    });


    assert.strictEqual(result.combos.length, 8);


    assert.strictEqual(result.combos[0].length, 40);
    assert.strictEqual(result.combos[0].maj_length, 12);
    assert.strictEqual(result.combos[0].min_length, 5);


    assert.strictEqual(result.combos[7].length, 50);
    assert.strictEqual(result.combos[7].maj_length, 14);
    assert.strictEqual(result.combos[7].min_length, 6);
});

test('SAMA batch full parameter sweep', () => {

    const close = new Float64Array(testData.close.slice(0, 60));

    const batchResult = wasm.sama_batch(close, {
        length_range: [45, 50, 5],
        maj_length_range: [13, 14, 1],
        min_length_range: [5, 5, 0]
    });


    assert.strictEqual(batchResult.combos.length, 4);
    assert.strictEqual(batchResult.rows, 4);
    assert.strictEqual(batchResult.cols, 60);
    assert.strictEqual(batchResult.values.length, 4 * 60);


    for (let combo = 0; combo < batchResult.combos.length; combo++) {
        const length = batchResult.combos[combo].length;
        const rowStart = combo * 60;
        const rowData = batchResult.values.slice(rowStart, rowStart + 60);



        for (let i = 0; i < length; i++) {
            assert(!isNaN(rowData[i]), `Should have valid value at index ${i} for length ${length}`);
        }


        for (let i = length; i < Math.min(length + 5, 60); i++) {
            assert(!isNaN(rowData[i]), `Unexpected NaN at index ${i} for length ${length}`);
        }
    }
});

test('SAMA batch edge cases', () => {

    const close = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);


    const singleBatch = wasm.sama_batch(close, {
        length_range: [10, 10, 1],
        maj_length_range: [5, 5, 0],
        min_length_range: [3, 3, 0]
    });

    assert.strictEqual(singleBatch.values.length, 15);
    assert.strictEqual(singleBatch.combos.length, 1);


    const largeBatch = wasm.sama_batch(close, {
        length_range: [10, 12, 10],
        maj_length_range: [5, 5, 0],
        min_length_range: [3, 3, 0]
    });


    assert.strictEqual(largeBatch.values.length, 15);
    assert.strictEqual(largeBatch.combos.length, 1);


    assert.throws(() => {
        wasm.sama_batch(new Float64Array([]), {
            length_range: [10, 10, 0],
            maj_length_range: [5, 5, 0],
            min_length_range: [3, 3, 0]
        });
    }, /All values are NaN|Input data slice is empty/);
});

test('SAMA batch vs single calculation', () => {

    const close = new Float64Array(testData.close.slice(0, 100));


    const singleResult = wasm.sama_js(close, 45, 13, 5);


    const batchResult = wasm.sama_batch(close, {
        length_range: [45, 45, 0],
        maj_length_range: [13, 13, 0],
        min_length_range: [5, 5, 0]
    });


    assertArrayClose(
        batchResult.values,
        singleResult,
        1e-10,
        "Batch vs single calculation mismatch"
    );
});


test('SAMA zero-copy API', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);
    const length = 10;
    const majLength = 5;
    const minLength = 3;


    const ptr = wasm.sama_alloc(data.length);
    assert(ptr !== 0, 'Failed to allocate memory');


    const memory = wasm.__wasm.memory;
    const memView = new Float64Array(
        memory.buffer,
        ptr,
        data.length
    );


    memView.set(data);


    try {
        wasm.sama_into(ptr, ptr, data.length, length, majLength, minLength);


        const regularResult = wasm.sama_js(data, length, majLength, minLength);
        for (let i = 0; i < data.length; i++) {
            if (isNaN(regularResult[i]) && isNaN(memView[i])) {
                continue;
            }
            assert(Math.abs(regularResult[i] - memView[i]) < 1e-10,
                   `Zero-copy mismatch at index ${i}: regular=${regularResult[i]}, zerocopy=${memView[i]}`);
        }
    } finally {

        wasm.sama_free(ptr, data.length);
    }
});

test('SAMA zero-copy with large dataset', () => {
    const size = 10000;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = Math.sin(i * 0.01) + Math.random() * 0.1;
    }

    const ptr = wasm.sama_alloc(size);
    assert(ptr !== 0, 'Failed to allocate large buffer');

    try {
        const memory = wasm.__wasm.memory;
        const memView = new Float64Array(memory.buffer, ptr, size);
        memView.set(data);

        wasm.sama_into(ptr, ptr, size, 50, 14, 6);


        const memory2 = wasm.__wasm.memory;
        const memView2 = new Float64Array(memory2.buffer, ptr, size);


        for (let i = 0; i < 50; i++) {
            assert(!isNaN(memView2[i]), `Should have valid value at index ${i}`);
        }


        for (let i = 50; i < Math.min(60, size); i++) {
            assert(!isNaN(memView2[i]), `Unexpected NaN at index ${i}`);
        }
    } finally {
        wasm.sama_free(ptr, size);
    }
});

test('SAMA batch_into zero-copy', () => {
    const size = 100;
    const data = new Float64Array(size);
    for (let i = 0; i < size; i++) {
        data[i] = 50 + Math.sin(i * 0.1) * 10;
    }


    const inPtr = wasm.sama_alloc(size);
    assert(inPtr !== 0, 'Failed to allocate input buffer');


    const memory = wasm.__wasm.memory;
    const inView = new Float64Array(memory.buffer, inPtr, size);
    inView.set(data);


    const numCombos = 2 * 2 * 2;
    const outPtr = wasm.sama_alloc(size * numCombos);
    assert(outPtr !== 0, 'Failed to allocate output buffer');

    try {

        const rowCount = wasm.sama_batch_into(
            inPtr, outPtr, size,
            40, 45, 5,
            12, 13, 1,
            4, 5, 1
        );

        assert.strictEqual(rowCount, numCombos, 'Should return correct row count');


        const memory2 = wasm.__wasm.memory;
        const outView = new Float64Array(memory2.buffer, outPtr, size * numCombos);


        const singleResult = wasm.sama_js(data, 40, 12, 4);
        const firstRow = Array.from(outView.slice(0, size));
        assertArrayClose(firstRow, singleResult, 1e-10, 'First row should match single calc');

    } finally {
        wasm.sama_free(inPtr, size);
        wasm.sama_free(outPtr, size * numCombos);
    }
});


test('SAMA SIMD128 consistency', () => {


    const testCases = [
        { size: 15, length: 10, majLength: 5, minLength: 3 },
        { size: 100, length: 50, majLength: 14, minLength: 6 },
        { size: 1000, length: 200, majLength: 20, minLength: 10 },
        { size: 5000, length: 500, majLength: 50, minLength: 25 }
    ];

    for (const testCase of testCases) {
        const data = new Float64Array(testCase.size);
        for (let i = 0; i < testCase.size; i++) {
            data[i] = 50 + Math.sin(i * 0.1) * 10 + Math.cos(i * 0.05) * 5;
        }

        const result = wasm.sama_js(data, testCase.length, testCase.majLength, testCase.minLength);


        assert.strictEqual(result.length, data.length);


        for (let i = 0; i < testCase.length; i++) {
            assert(!isNaN(result[i]), `Should have valid value at index ${i} for size=${testCase.size}`);
        }


        let sumAfterWarmup = 0;
        let countAfterWarmup = 0;
        for (let i = testCase.length; i < result.length; i++) {
            assert(!isNaN(result[i]), `Unexpected NaN at index ${i} for size=${testCase.size}`);
            sumAfterWarmup += result[i];
            countAfterWarmup++;
        }


        if (countAfterWarmup > 0) {
            const avgAfterWarmup = sumAfterWarmup / countAfterWarmup;
            assert(avgAfterWarmup > 0, `Average value ${avgAfterWarmup} should be positive`);
            assert(avgAfterWarmup < 100, `Average value ${avgAfterWarmup} seems too large`);
        }
    }
});


test('SAMA zero-copy error handling', () => {

    assert.throws(() => {
        wasm.sama_into(0, 0, 10, 50, 14, 6);
    }, /null pointer|invalid memory/i);


    const ptr = wasm.sama_alloc(10);
    try {

        assert.throws(() => {
            wasm.sama_into(ptr, ptr, 10, 0, 14, 6);
        }, /Invalid period/);


        assert.throws(() => {
            wasm.sama_into(ptr, ptr, 10, 10, 0, 6);
        }, /Invalid period/);


        assert.throws(() => {
            wasm.sama_into(ptr, ptr, 10, 10, 14, 0);
        }, /Invalid period/);
    } finally {
        wasm.sama_free(ptr, 10);
    }
});


test('SAMA zero-copy memory management', () => {

    const sizes = [100, 1000, 10000];

    for (const size of sizes) {
        const ptr = wasm.sama_alloc(size);
        assert(ptr !== 0, `Failed to allocate ${size} elements`);


        const memory = wasm.__wasm.memory;
        const memView = new Float64Array(memory.buffer, ptr, size);
        for (let i = 0; i < Math.min(10, size); i++) {
            memView[i] = i * 1.5;
        }


        for (let i = 0; i < Math.min(10, size); i++) {
            assert.strictEqual(memView[i], i * 1.5, `Memory corruption at index ${i}`);
        }


        wasm.sama_free(ptr, size);
    }
});

test('SAMA batch error handling', () => {
    const close = new Float64Array(testData.close.slice(0, 10));


    assert.throws(() => {
        wasm.sama_batch(close, {
            length_range: [40, 40],
            maj_length_range: [12, 12, 0],
            min_length_range: [5, 5, 0]
        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.sama_batch(close, {
            length_range: [40, 40, 0],
            maj_length_range: [12, 12, 0]

        });
    }, /Invalid config/);


    assert.throws(() => {
        wasm.sama_batch(close, {
            length_range: "invalid",
            maj_length_range: [12, 12, 0],
            min_length_range: [5, 5, 0]
        });
    }, /Invalid config/);
});


test('SAMA wasm() function compatibility', () => {
    const data = new Float64Array([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]);


    const result1 = wasm.sama_js(data, 10, 5, 3);
    assert.strictEqual(result1.length, data.length);


    const result2 = wasm.sama_js(data, 5, 3, 2);
    assert.strictEqual(result2.length, data.length);


    const result3 = wasm.sama_js(data, 14, 7, 4);
    assert.strictEqual(result3.length, data.length);
});

test.after(() => {
    console.log('SAMA WASM tests completed');
});
